internal/internal_deployment_client.go

```diff
@@ -0,0 +1,313 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.temporal.io/api/common/v1"
+	"go.temporal.io/api/deployment/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/sdk/converter"
+)
+
+type (
+	// deploymentClient is the client for managing deployments.
+	deploymentClient struct {
+		workflowClient *WorkflowClient
+	}
+
+	// deploymentListIteratorImpl is the implementation of [DeploymentListIterator].
+	// Adapted from [scheduleListIteratorImpl].
+	deploymentListIteratorImpl struct {
+		// nextDeploymentIndex - Local index to cached deployments.
+		nextDeploymentIndex int
+
+		// err - Error from getting the last page of deployments.
+		err error
+
+		// response - Last page of deployments from server.
+		response *workflowservice.ListDeploymentsResponse
+
+		// paginate - Function to get the next page of deployment from server.
+		paginate func(nexttoken []byte) (*workflowservice.ListDeploymentsResponse, error)
+	}
+)
+
+func (iter *deploymentListIteratorImpl) HasNext() bool {
+	if iter.err == nil &&
+		(iter.response == nil ||
+			(iter.nextDeploymentIndex >= len(iter.response.Deployments) && len(iter.response.NextPageToken) > 0)) {
```

Comment:
`ListDeployments` uses `visibility.ListWorkflowExecutions` under the hood directly, the same way that `ListSchedules` does, so the contents of a page and the `nextPageToken` are handled the same for `ListDeployments` and `ListSchedules`. Because of this, I believe that there is no risk of an empty page followed by a non-empty page.

Reply:
In the elastic search visibility store implementation, next page token is only non-nil if the number of entries on the page == pageSize: https://github.com/temporalio/temporal/blob/main/common/persistence/visibility/store/elasticsearch/visibility_store.go#L839-L848

Reply:
@cretz @antlai-temporal hope this helps

Reply:
same in the sql visibility store implementation, next page token is only non-nil if the number of entries on the page == pageSize: https://github.com/temporalio/temporal/blob/main/common/persistence/visibility/store/sql/visibility_store.go#L233-L248

Reply:
> I believe that there is no risk of an empty page followed by a non-empty page

Sounds good. Just wanted to make sure there was no post-visibility item filtering done server side that would affect this.